### PR TITLE
Add variable conditions for Rainbow Bridge and Ganon's BK on LACS

### PR DESCRIPTION
**Pull request by Roman971**
Both the rainbow bridge and Ganon BK settings are now bound to various sliders that allow customizing the exact count of stones, medallions, dungeons or tokens required. This also includes a new Ganon BK condition on LACS with tokens, for consistency/balance with the bridge setting (credits to @Cuphat for that part).

Currently, none of the sliders are randomized with "Randomize Main Rules", and the LACS token option is excluded from randomization as well, just like the bridge token option.

Conditional always hints are updated accordingly, mostly based on the "rule" that locations should have always hints if they require 2 or more dungeon rewards in the worst case scenario.

### CODE REVIEW

**Comment by Roman971** [comment]
Noting here so we don't forget but this will require small updates (already known and tested) to fit with PRs #1102 and #1112 if they get merged before. Otherwise they will have to be updated accordingly if this one gets merged first. Either order is fine tho.

File: State.py [hunk 1]
```diff
@@ -60,6 +60,10 @@ def has_all_of(self, items):
         return all(map(self.prog_items.__contains__, items))
 
 
+    def count_of(self, items):
+        return len(list(filter(self.prog_items.__contains__, items)))
```
Comment (Zannick) [inline State.py:64]: `sum(map` instead of `len(list(filter`

File: World.py [hunk 2]
```diff
@@ -179,11 +181,13 @@ def __init__(self, id, settings):
...
-            max_tokens = self.bridge_tokens
+            max_tokens = max(max_tokens, self.bridge_tokens)
+        if self.lacs_condition == 'tokens':
+            max_tokens = max(max_tokens, self.lacs_tokens)
         tokens = [50, 40, 30, 20, 10]
         for t in tokens:
-            if t > max_tokens and f'{t} Gold Skulltula Reward' not in self.disabled_locations:
-                max_tokens = t
+            if f'{t} Gold Skulltula Reward' not in self.disabled_locations:
+                max_tokens = max(max_tokens, t)
```
Comment (Zannick) [inline World.py:190]: This doesn't look to me like a functional difference, except for checking each of the skulltula rewards prior to checking if `t > max_tokens` (in max). The other way used the `t > max_tokens` check as a short-circuit to skip the other check.

File: SettingsList.py [hunk 3]
```diff
@@ -1770,26 +1770,27 @@ def __init__(self, name, gui_text, min, max, default, step=1,
...
+            'dungeons':   'Dungeons',
             'tokens':     'Gold Skulltula Tokens'
         },
         gui_tooltip    = '''\
             'Always Open': Rainbow Bridge is always present.
             'Vanilla Requirements': Spirit/Shadow Medallions and Light Arrows.
-            'All Spiritual Stones': All 3 Spiritual Stones.
-            'All Medallions': All 6 Medallions.
-            'All Dungeons': All Medallions and Spiritual Stones.
+            'Spiritual Stones': A configurable amount of 3 Spiritual Stones.
```
Comment (Zannick) [inline SettingsList.py:1781]: Nit on the wording here: "amount of 3" feels awkward; would "A configurable amount, up to 3..." make sense?
Or, LACS says "A configurable amount of Spiritual Stones." omitting the number which I think is fine.

**Review by Zannick — CHANGES_REQUESTED** [review]

File: State.py (hunk 1 above)
Comment (Roman971) [reply State.py:64]: Isn't `sum` intended only for numeric values? I feel like using it on a map of booleans might work in current versions of Python but could very well break later.

File: World.py (hunk 2 above)
Comment (Roman971) [reply World.py:190]: I believe this was done for consistency with the other cases, but @Cuphat actually made these changes, so he's probably more fit to answer this.

Personally I don't mind either patterns, tho I'd argue consistency is better than a slight performance improvement in this case.

Comment (Cuphat) [reply World.py:190]: Yeah, it was just consistency while I was changing all the others and adding Ganon BK from tokens. I don't have a strong opinion on that particular change.

File: State.py (hunk 1 above)
Comment (Zannick) [reply State.py:64]: `bool` is a numeric type, but you're right that it might break later.

**Review by Zannick — APPROVED** [review]

**Review by Zannick — APPROVED** [review]

**Pull request merged**